### ISSUE TITLE
Bug with array payload validation

### ISSUE THREAD

**Issue by alex-zywicki** [dedoussis/asynction#148]
When using:
``` yaml
payload:
  type: array
```
And sending any array to to the server I get this error.

```
Traceback (most recent call last):
 File "/usr/local/lib/python3.9/site-packages/asynction/validation.py", line 26, in jsonschema_validate_with_error_handling
   jsonschema.validate(instance, schema)
 File "/usr/local/lib/python3.9/site-packages/jsonschema/validators.py", line 967, in validate
   raise error
jsonschema.exceptions.ValidationError: ([],) is not of type 'array'
Failed validating 'type' in schema:
   {'type': 'array'}
On instance:
   ([],)
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
 File "src/gevent/greenlet.py", line 906, in gevent._gevent_cgreenlet.Greenlet.run
 File "/usr/local/lib/python3.9/site-packages/socketio/server.py", line 720, in _handle_event_internal
   r = server._trigger_event(data[0], namespace, sid, *data[1:])
 File "/usr/local/lib/python3.9/site-packages/socketio/server.py", line 745, in _trigger_event
   return self.handlers[namespace][event](*args)
 File "/usr/local/lib/python3.9/site-packages/flask_socketio/__init__.py", line 282, in _handler
   return self._handle_event(handler, message, namespace, sid,
 File "/usr/local/lib/python3.9/site-packages/flask_socketio/__init__.py", line 766, in _handle_event
   ret = handler(*args)
 File "/usr/local/lib/python3.9/site-packages/asynction/validation.py", line 97, in handler_with_validation
   validate_payload(args, message.payload)
 File "/usr/local/lib/python3.9/site-packages/asynction/validation.py", line 57, in validate_payload
   jsonschema_validate_payload(args, schema)
 File "/usr/local/lib/python3.9/site-packages/asynction/validation.py", line 28, in jsonschema_validate_with_error_handling
   raise exc_type(str(e))
asynction.exceptions.PayloadValidationException: ([],) is not of type 'array'
Failed validating 'type' in schema:
   {'type': 'array'}
On instance:
   ([],)
2021-12-15T20:29:58Z <Thread at 0x7feeac4cc480: <bound method Server._handle_event_internal of <socketio.server.Server object at 0x7feead801700>>(<socketio.server.Server object at 0x7feead801700>, '0Fn-hKNPKV47OogIAAAF', 'HR_USU2XamBt4DyHAAAE', ['subscriptions', []], '/subscriptions/', 3)> failed with PayloadValidationException
```

After Some quick digging I tried changing:
``` python3
def validate_payload(args: Sequence, schema: Optional[JSONSchema]) -> None:
    if schema is None:
        if args:
            raise PayloadValidationException(
                "Handler args provided for message that has no payload defined."
            )
        # Validation succeeded since there is no message payload specified
        # and no args have been provided.
        return

    schema_type = schema["type"]
    if schema_type == "array":
        jsonschema_validate_payload(args, schema)
    else:
        if len(args) > 1:
            raise PayloadValidationException(
                "Multiple handler arguments provided, "
                f"although schema type is: {schema_type}"
            )

        jsonschema_validate_payload(args[0], schema)
```

to
``` python3
def validate_payload(args: Sequence, schema: Optional[JSONSchema]) -> None:
    if schema is None:
        if args:
            raise PayloadValidationException(
                "Handler args provided for message that has no payload defined."
            )
        # Validation succeeded since there is no message payload specified
        # and no args have been provided.
        return

    schema_type = schema["type"]
    if schema_type == "array":
        jsonschema_validate_payload(args[0], schema)  # added the [0] here
    else:
        if len(args) > 1:
            raise PayloadValidationException(
                "Multiple handler arguments provided, "
                f"although schema type is: {schema_type}"
            )

        jsonschema_validate_payload(args[0], schema)
```

And the error was no longer present.

Though this does seem to go against the logic in place for the special case provided there.
Why is this special case there?
This issue only seems to occur when the top level payload object type is an array.

it appears that the validator is receiving the `args` as a length 1 tuple where the 0th element is the actual payload.

The else case seems to handle that, but the "array" case seems to assume that the `args` is the array itself

**Comment by alex-zywicki** [comment]
@dedoussis If I am interpreting things correctly I think that the special case for the `array` is incorrect and can be removed. Or that some checking needs to be done against the `args` parameter.

**Comment by dedoussis** [comment]
This condition branching is there to account for the case of the handler accepting multiple args. In the case of multiple args, the payload scheme is an tuple array. See the payload structure table in the middle of this section of my article: https://www.asyncapi.com/blog/socketio-part2#channels

In the case of an tuple payload, all args need to be validated and not just the first one.
However, I think I have not accounted for the case of the payload being an array rather than an tuple. In the case of an array, we will only need to validate the first arg. Hence we need to have an extra nested condition within `if schema_type == "array"` to check if there is a `prefixItems` field or not.

**Comment by alex-zywicki** [comment]
@dedoussis I guess what I don't understand is why in the one case where the payload is an `array` but does not have `prefixItems` that `args would be a tuple containing an array, but in the `prefixItems` case that args itself would be the array?

Why would:
```yaml
type: "array"
```
Result in `([1,2,3], )` as `args`

But
```yaml
type: "array"
prefixItems:
    ...
```
Result in `(1,2,3)`

My interpretation of the second case there is that there were multiple payloads rather than an array payload. from everything I've seen (using python-socketio as the client and asynction) the `args` on a message handler will always only have one item (the payload)

**Comment by dedoussis** [comment]
```yaml
type: "array"
items:
  type: number
```
The above payload signals that there should only be a single argument of type `Sequence[str]`, ie `args == ([1,2,3], )`.
Using python-socketio as the client, an event emission of such a payload would look like this:
```python
c.emit("foo-event", data=[1,2,3])
```
Under the hood, the `python-socketio` client does some magic to convert non-tuple data to a tuple of a single element, so that `[1,2,3]` becomes `([1,2,3], )`.

And in asynction the handler would look like this:
```python
def on_foo_event(nums: Sequence[int]) -> None:
    ...
```

Whereas:
```yaml
type: "array"
prefixItems:
- type: string
- type: number
```
signals that there should be 2 arguments (one of type string and one of type int, ie `args == ("bar", 1)`)

Using python-socketio as the client, an event emission of such a payload would look like this:
```python
c.emit("foo-event", data=("bar", 1))
```
Here we pass a tuple as data, so python-socketio won't do any magic conversion.

And in asynction the handler would look like this:
```python
def on_foo_event(word: str, num: int) -> None:
    ...
```

**Comment by dedoussis** [comment]
This fix has been released as part of version `0.7.0`. @alex-zywicki If you're happy with it, I can close the issue.

**Comment by alex-zywicki** [comment]
@dedoussis close away
